Add recommended security tools document (#25)
redhat-appstudio/book: tools/security-tools.MD

# Security Guild recommended tools

> This document lists the tools that are available and researched for performing linting, Static Code Analysis, vulnerability scanning on different platforms and languages. The goal is to introduce one of more of these tools into StoneSoup.

Reviewer: ```suggestion
This document lists the available tools that we researched for performing linting, Static Code Analysis, and vulnerability scanning on different platforms and languages. The goal is to introduce one of more of these tools into StoneSoup for customers to use.
```

> From our discussions in the Secure Engineering Guild, our current position is that we would encourage teams to use the tools highlighted in this list for consistency. But if a team finds a different tool that's a better fit-for-purpose, that is also fine. The ProdSec org also supports some tools, and some tools will be integrated into StoneSoup over time (taking the responsibility of implementation off of our teams). So where we have a recommendation, there it also makes sense to follow it if it fits your team/product's needs.  

Reviewer: ```suggestion
From our discussions in the Secure Engineering Guild, our current position is that we would encourage teams to use the tools highlighted in this list for consistency. But if a team finds a different tool that's a better fit-for-purpose, that is also fine. The ProdSec org also supports some tools, and some tools will be integrated into our product over time (taking the responsibility of implementation off of our teams). Where we have a recommendation, it makes sense to follow it if it fits your team/product's needs.  
```

> ProdSec recommends us to not use tools that could potentially "phone home" or any that could otherwise expose information about embargoed vulnerabilities unless they have previously approved our particular use of the tool.

Reviewer: ```suggestion
Per our discussions with ProdSec, we should not use tools that could potentially "phone home," or any that could otherwise expose information about embargoed vulnerabilities, unless they have previously approved our particular use of the tool.
```

### Linting Tools

#### Golang
**golangci-lint** - https://github.com/golangci/golangci-lint

```
Usage:
$ cd my/golang/project
$ golangci-lint run
```

#### Docker
**hadolint** - https://github.com/hadolint/hadolint  
_Note: In order to lint Bash code inside a Dockerfile, it uses Shellcheck internally._  
  
```
Usage:
$ hadolint Dockerfile  
(or)  
$ docker run --rm -i ghcr.io/hadolint/hadolint < Dockerfile
```

### Vulnerability Scanners

**clair** - https://github.com/quay/clair  
- quay.io uses Clair internally and the project is officially under them. 

> - Check these [docs](https://quay.github.io/clair/howto/deployment.html) to understand the deployment models Clair currently uses which also talks about on setting it up in an OpenShift or a Kubernetes environment. 

Reviewer: I recommend dropping "which also talks about on setting it up in an OpenShift or a Kubernetes environment". Otherwise make this a couple of sentences for readability.

> - Check these [docs](https://quay.github.io/clair/howto/deployment.html) to understand the deployment models Clair currently uses which also talks about on setting it up in an OpenShift or a Kubernetes environment. 

Reviewer: ```suggestion
- Check these [docs](https://quay.github.io/clair/howto/deployment.html) to understand the deployment models Clair currently uses. 
```

- For teams using quay.io as their container image registry, we enjoy the benefit of these scans via their website. You can check the results under the vulnerabilites tab of an image.

> Note: [clair-in-ci](https://quay.io/repository/redhat-appstudio/clair-in-ci) is a feature which includes security scanning via clair by default for any Pipelines created in StoneSoup.

Reviewer: ```suggestion
_Note: [clair-in-ci](https://quay.io/repository/redhat-appstudio/clair-in-ci) is a feature which includes security scanning via clair. It is enabled by default for any Pipelines created in StoneSoup._
```

Reviewer: Is there a Tekton task link for this one? It's nice that you included it for the others.

Author: Ah, missed the link on this one. Updated it now.

> _Note: [clair-in-ci](https://quay.io/repository/redhat-appstudio/clair-in-ci) is a feature which includes security scanning via clair. It is enabled by default for any Pipelines created in StoneSoup. A Tekton Task is available that can be used to run gosec in your own Pipelines [here](https://github.com/redhat-appstudio/build-definitions/blob/main/task/get-clair-scan/)._

Reviewer: Thanks for adding this. Just two more requests:
1) Change this to Clair instead of gosec: "can be used to run gosec in your own Pipelines"
2) The link isn't working for me. I get a 404 not found. Is it https://github.com/redhat-appstudio/build-definitions/tree/main/task/clair-scan/ instead?

Author: Both points updated now. Thanks for catching the link issue.

### SAST Tools

**_gosec_** - https://github.com/securego/gosec
 
_Note: StoneSoup uses gosec to perform static analysis on our Golang projects. A Tekton Task is available that can be used to run gosec in your own Pipelines [here](https://github.com/redhat-appstudio/build-definitions/blob/main/task/get-clair-scan/)._

**find-sec-bugs** - https://github.com/find-sec-bugs/find-sec-bugs

> _Note: StoneSoup uses find-sec-bug to perform static analysis on our Java builds. A Tekton Task is available that can be used to run find-sec-bugs on a Java application in your own Pipelines [here](https://github.com/redhat-appstudio/build-definitions/blob/main/task/sast-java-sec-check)._

Reviewer: ```suggestion
_Note: StoneSoup uses find-sec-bugs to perform static analysis on our Java builds. A Tekton Task is available that can be used to run find-sec-bugs on a Java application in your own Pipelines [here](https://github.com/redhat-appstudio/build-definitions/blob/main/task/sast-java-sec-check)._
```
